### ISSUE TITLE
Support System.Collections.Immutable types

### ISSUE THREAD

**Issue by latonz** [riok/mapperly#250]
Support collection types in `System.Collections.Immutable` as mapping target types.
Eg. Mapperly should be able to implement a mapping from an `IEnumerable<long>` to an `ImmutableList<int>`.

**Comment by jamesfoster** [comment]
While this feature is not implemented is there a generic way to work around this?

Something like:

```csharp
private static ImmutableList<TDest> ToImmutableList<TDest, TSource>(
    IEnumberable<TSource> source,
    Func<TSource, TDest> map
)
{
   return source.Select(map).ToImmutableList();
}
```

We can obviously add a workaround for each type but it doesn't seem to be possible to implement it generically.

**Comment by latonz** [comment]
Currently there is no support for generic user implemented mapping methods... I can't think of a feasible workaround 😢

**Comment by jamesfoster** [comment]
Thanks for the response. Without knowing too much about how source generators work, would it not be possible to support the suggested code? This could be useful for any generic type with a single type argument. if you know how to map from `TSource` to `TDest` you can supply a value for `map`. You can then use this method to map from any `A<TSource>` to a `B<TDest>`.

**Comment by TimothyMakkison** [comment]
> Currently there is no support for generic user implemented mapping methods... I can't think of a feasible workaround 😢

I'll give it a go. Looks like `EnumerableMappingBuilder.ResolveCollectMethodName` can be used to add `ToImmutableList` etc. Not sure which immutable collections should be added.

**Comment by latonz** [comment]
@jamesfoster this could probably be a new feature of Mapperly. However, I think it adds a lot of complexity and I'm not sure if it justifies the added complexity.

**Comment by jamesfoster** [comment]
> @jamesfoster this could probably be a new feature of Mapperly. However, I think it adds a lot of complexity and I'm not sure if it justifies the added complexity.

No worries. My use case was immutable collections anyway, and it looks like you're working on it. I can work around it for the time being by adding the non-generic mapper for each type as needed.

Thanks.

**Comment by latonz** [comment]
https://github.com/riok/mapperly/pull/305 implements support for the following immutable collections:
* `ImmutableList<T>`
* `ImmutableArray<T>`
* `ImmutableHashSet<T>`
* `ImmutableQueue<T>`
* `ImmutableSortedSet<T>`

**Comment by latonz** [comment]
Thanks to another great contribution from @TimothyMakkison  `ImmutableDictionary<TKey, TValue>` and `ImmutableSortedDictionary<TKey,TValue>` are now supported too.

**Comment by latonz** [comment]
This issue has been released in version 2.8.0-next.2 which is available on nuget.
